forch/faucetizer.py

```diff
@@ -1,49 +1,182 @@
 """Collect Faucet information and generate ACLs"""
 
+import argparse
+import copy
 import logging
 import os
 import sys
+import yaml
 
 from forch.forchestrator import configure_logging
-from forch.utils import proto_dict
 from forch.utils import yaml_proto
 
 from forch.proto.network_state_pb2 import NetworkState
 
-LOGGER = logging.getLogger('topology')
+LOGGER = logging.getLogger('faucetizer')
+
+
+class Device:
+    """A device with learning and behavior"""
+    def __init__(self):
+        self.old_learning = None
+        self.new_learning = None
+        self.behavior = None
+
+    def has_migrated(self):
+        """Determine if the switch or port of a device has changed"""
+        if not self.old_learning:
+            return False
+        is_same_switch = self.old_learning.switch != self.new_learning.switch
+        is_same_port = self.old_learning.port != self.new_learning.port
+        return not is_same_switch or not is_same_port
+
+    def is_fulfilled(self):
+        """Determine if the info of device is enough to update output faucet config"""
+        return self.new_learning and self.behavior
+
+    def commit(self):
+        self.old_learning = copy.deepcopy(self.new_learning)
 
 
 class Faucetizer:
     """Collect Faucet information and generate ACLs"""
-    def __init__(self):
-        self._network_state = {}
+    def __init__(self, output_file):
+        self._devices = {}
+        self._base_faucet_config = None
+        self._faucet_config = None
+        self._output_file = output_file
 
     def process_network_state(self, network_state):
-        """Process network state inputs"""
-        self._network_state = proto_dict(network_state, True)
+        """Process network state input"""
+        macs = set()
+        for mac, learning in network_state.device_mac_learnings.items():
+            macs.add(mac)
+            self._process_device_learning(mac, learning)
+        for mac, behavior in network_state.device_mac_behaviors.items():
+            macs.add(mac)
+            self._process_device_behavior(mac, behavior)
+
+        self._faucetize(macs)
+
+    def _process_device_learning(self, mac, learning):
+        device = self._devices.setdefault(mac, Device())
+        device.new_learning = learning
+
+    def _process_device_behavior(self, mac, behavior):
+        device = self._devices.setdefault(mac, Device())
+        device.behavior = behavior
+
+    def process_faucet_config(self, faucet_config):
+        """Process faucet config when base faucet config changes"""
+        self._base_faucet_config = faucet_config
+        self._faucet_config = faucet_config
+        self._faucetize()
+
+    def _faucetize(self, macs=None):
+        if not self._faucet_config:
```

Comment:
this should be depenedent on the base config existing, not a previous output config, and every time through this function, it should start with a clean (deep) copy of _base_faucet_config

Reply:
I think it should be enough to start with a copy of self._out_faucet_config if we use a local variable in the following processing? This way devices map will be consitent with self._out_faucet_config.

Reply:
I suppose -- for now it's fine as long as it works. It feels like you're over-optimizing and likely to create corner cases (e.g. state left-over in the out faucet config that isn't accurately reflected in the current setup), which are much harder to deal with than clean logic (even if it's slower).

What happens, for example, if/when the base faucet config changes?

Reply:
Also, I don't see where you "start with a copy of" -- did the copy get missed?